### ISSUE TITLE
unexpected error error="uninitialized constant HTTP"

### ISSUE THREAD

**Issue by quasipedia** [fluent/fluentd#539]
I can't manage to run `fluentd` on Fedora 21.  A few info on my system:
```
--- ~ » ruby --version
ruby 2.1.5p273 (2014-11-13 revision 48405) [x86_64-linux]
--- ~ » fluentd --version
fluentd 0.12.2
```
Basically what happens when I try to run it, is:
```
--- ~ » fluentd -c dev/fluent.conf 
2015-01-12 14:39:18 +0100 [info]: reading config file path="dev/fluent.conf"
2015-01-12 14:39:18 +0100 [info]: starting fluentd-0.12.2
2015-01-12 14:39:18 +0100 [error]: unexpected error error="uninitialized constant HTTP"
  2015-01-12 14:39:18 +0100 [error]: /usr/local/share/gems/gems/http_parser.rb-0.6.0/lib/http_parser.rb:4:in `<top (required)>'
  2015-01-12 14:39:18 +0100 [error]: /usr/local/share/ruby/site_ruby/rubygems/core_ext/kernel_require.rb:54:in `require'
  2015-01-12 14:39:18 +0100 [error]: /usr/local/share/ruby/site_ruby/rubygems/core_ext/kernel_require.rb:54:in `require'
  2015-01-12 14:39:18 +0100 [error]: /usr/local/share/gems/gems/http_parser.rb-0.6.0/lib/http/parser.rb:1:in `<top (required)>'
  2015-01-12 14:39:18 +0100 [error]: /usr/local/share/ruby/site_ruby/rubygems/core_ext/kernel_require.rb:54:in `require'
  2015-01-12 14:39:18 +0100 [error]: /usr/local/share/ruby/site_ruby/rubygems/core_ext/kernel_require.rb:54:in `require'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/plugin/in_http.rb:23:in `<class:HttpInput>'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/plugin/in_http.rb:18:in `<module:Fluent>'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/plugin/in_http.rb:17:in `<top (required)>'
  2015-01-12 14:39:18 +0100 [error]: /usr/local/share/ruby/site_ruby/rubygems/core_ext/kernel_require.rb:54:in `require'
  2015-01-12 14:39:18 +0100 [error]: /usr/local/share/ruby/site_ruby/rubygems/core_ext/kernel_require.rb:54:in `require'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/plugin.rb:85:in `block in load_plugin_dir'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/plugin.rb:83:in `each'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/plugin.rb:83:in `load_plugin_dir'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/plugin.rb:78:in `load_plugins'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/engine.rb:43:in `init'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/supervisor.rb:415:in `init_engine'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/supervisor.rb:136:in `block in start'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/supervisor.rb:266:in `call'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/supervisor.rb:266:in `main_process'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/supervisor.rb:241:in `block in supervise'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/supervisor.rb:240:in `fork'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/supervisor.rb:240:in `supervise'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/supervisor.rb:134:in `start'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/lib/fluent/command/fluentd.rb:167:in `<top (required)>'
  2015-01-12 14:39:18 +0100 [error]: /usr/local/share/ruby/site_ruby/rubygems/core_ext/kernel_require.rb:69:in `require'
  2015-01-12 14:39:18 +0100 [error]: /usr/local/share/ruby/site_ruby/rubygems/core_ext/kernel_require.rb:69:in `require'
  2015-01-12 14:39:18 +0100 [error]: /home/mac/.gem/ruby/gems/fluentd-0.12.2/bin/fluentd:6:in `<top (required)>'
  2015-01-12 14:39:18 +0100 [error]: /usr/local/bin/fluentd:23:in `load'
  2015-01-12 14:39:18 +0100 [error]: /usr/local/bin/fluentd:23:in `<main>'
2015-01-12 14:39:18 +0100 [info]: process finished code=256
2015-01-12 14:39:18 +0100 [warn]: process died within 1 second. exit.
```
The problem began when I upgraded my system, which required me to both upgrade Fedora and all gems with `gem update` and `gem update --system`.  Prior to that, the system worked just fine.

For completeness, this is my `fluent.conf` file:

```
<source>
  type forward
</source>
<match **>
  type elasticsearch
  buffer_type memory
  flush_interval 1s
  retry_wait 1s
  disable_retry_limit true
  max_retry_wait 1h
  logstash_format true
</match>

```

**Comment by repeatedly** [comment]
> 2015-01-12 14:39:18 +0100 [error]: /usr/local/share/gems/gems/
http_parser.rb-0.6.0/lib/http_parser.rb:4:in `<top (required)>'

From this line, there is no HTTP module in your gem.

https://github.com/tmm1/http_parser.rb/blob/10dd2edf301fa8d1639cab7ed1c152c659866d51/lib/http_parser.rb#L4
https://github.com/tmm1/http_parser.rb/blob/10dd2edf301fa8d1639cab7ed1c152c659866d51/ext/ruby_http_parser/ruby_http_parser.c#L468

I don't know what happen in your system but how about re-instanll
http_parser.rb gem?


Masahiro

**Comment by quasipedia** [comment]
@repeatedly - Thank you for the quick reply.  I have zero ruby experience and I may have messed up stuff perhaps.

When problem started to arise, I ran `sudo gem uninstall fluentd` and then re-installed it.  What happened is that it did not install _any_ of the dependencies (despite my gem being version `2.4.5`, where this is the default behaviour).

So I proceeded to install them manually, by installing a gem every time the interpreter complained about a missing file.  One thing that reading your comment made me realise, is that coming from the python world I have assumed `.rb` to be just the file ending, so I ran the install command without that `gem install http_parser` instead of `gem install http_parser.rb`.

Since nothing worked, I tried to install them in my userspace (no `sudo`), and here is where I believe the problem with the double locations may have sprung into existence.

Anyhow: thanks to your insight, I proceeded to wipe a few directories and fresh-install everything, and now it works again. Thanks a lot! :)
